3.12/interpreter_definition.md

```diff
@@ -0,0 +1,288 @@
+set of instructions and potentially different dispatching mechanism.
+But the instructions it will interpret will be built from the same building
+blocks as the instructions for the tier 1 (PEP 659) interpreter.
+
+Rewriting all the instructions is tedious and error-prone, and changing the
+instructions is a maintenance headache as both versions need to be kept in sync.
+
+By using a code generator and using a common source for the instructions, or 
+parts of instructions, we can reduce the potential for errors considerably.
+
+
+## Specification
+
+> This specification is at an early stage and is likely to change considerably.
+
+Syntax
+------
+
+Each op definition has a kind, a name, a stack and instruction stream effect,
+and a piece of C code describing its semantics::
+ 
+```
+  definition:
+    KIND NAME "(" (input ("," input))? "--" (output ("," output)*)? ")" "{" C-code "}"
+    |
+    KIND NAME "=" ( NAME | stream)+ ";"
+    |
+    KIND NAME "{" C-code "}"
+ 
+  input:
+    object | stream | array
+
+  output:
+    object | array
+
+  object:
+    NAME [":" NAME] [ "if" "(" C-expression ")" ]
+
+  stream:
+    (# | ## | ####) NAME
+
+  array:
+    object "[" NAME "]"
+
+  family:
+    FAMILY NAME = NAME ("," NAME)+ ";"
+
+   file:
+      (definition | family)+
+```
+
+The `KIND` must be one of:
+
+* `inst`: A normal instruction, as currently defined by `TARGET(NAME)` in `ceval.c`.
+* `op`: A part instruction from which other ops and instructions can be constructed.
+* `super`: A super-instruction, such as `LOAD_FAST__LOAD_FAST`.
+
+`NAME` can be any ASCII identifier that is a C identifier and not a C or Python keyword.
+`foo_1` is legal. `$` is not legal, nor is `struct` or `class`.
+
+The optional second name in an `object` is the type. It defaults to `PyObject *`.
+The objects before the "--" are the objects on the stack at the start of the instruction.
+Those after the "--" are the objects on the stack at the end of the instruction.
+
+The third form of `definition` is a transition form to allow the original C code 
+definitions to be copied. It lacks information to generate anything other than the
+interpreter, but is useful for initial porting of code.
+
+The number of `#`s in a `stream` define how many codeunits are consumed from the
+instruction stream.
+
+The name `oparg` is pre-defined as a 32 bit value fetched from the instruction stream.
+
+The C code may include special functions that are understood by the tools as
+part of the DSL.
+
+Those functions are:
+
+* `DEOPT_IF(cond)`. Deoptimize if `cond` is met.
+* `ERROR_IF(cond, handler)`. Jump to error handler if `cond` is true.
+* `PEEK(n)`. The `n`<sup>th</sup> item on the stack, which is not pushed.
+
+Variables can either be defined in the input, ouput, or in the C code.
+Variables defined in the input may not be assigned in the C code.
+If an `ERROR_IF` occurs, all values will be removed from the stack.
```

Comment:
That's up to the definition, it is part of the semantics.
Which means that the code must be written so that the refcounts are correct at `ERROR_IF` and `DEOPT_IF` points.